src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/Cohost/CohostOnTypeFormattingEndpoint.cs

```diff
@@ -0,0 +1,176 @@
+        // trigger characters are a superset of the C# triggers, so we can't use that as a sign. Secondly, whilst we
+        // might be making one Html request, it could be then calling CSS or TypeScript servers, so our single request
+        // to OOP could potentially save a few requests downstream. Lastly, our request to OOP is MessagePack which is
+        // generally faster than Json anyway.
+        var triggerKind = await _remoteServiceInvoker.TryInvokeAsync<IRemoteFormattingService, IRemoteFormattingService.TriggerKind>(
+            razorDocument.Project.Solution,
+            (service, solutionInfo, cancellationToken) => service.GetOnTypeFormattingTriggerKindAsync(solutionInfo, razorDocument.Id, request.Position.ToLinePosition(), request.Character, cancellationToken),
+            cancellationToken).ConfigureAwait(false);
+
+        if (triggerKind == IRemoteFormattingService.TriggerKind.Invalid)
+        {
+            return null;
+        }
+
+        var sourceText = await razorDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+        ImmutableArray<TextChange> htmlChanges = [];
+        if (triggerKind == IRemoteFormattingService.TriggerKind.ValidHtml)
+        {
+            _logger.LogDebug($"Getting Html formatting changes for {razorDocument.FilePath}");
+            var htmlResult = await GetHtmlFormattingEditsAsync(request, razorDocument, cancellationToken).ConfigureAwait(false);
+
+            if (htmlResult is not { } htmlEdits)
+            {
+                // We prefer to return null, so the client will try again
+                _logger.LogDebug($"Didn't get any edits back from Html");
+                return null;
+            }
+
+            htmlChanges = htmlEdits.SelectAsArray(sourceText.GetTextChange);
+        }
+
+        var options = RazorFormattingOptions.From(request.Options, clientSettings.AdvancedSettings.CodeBlockBraceOnNextLine);
+
+        _logger.LogDebug($"Calling OOP with the {htmlChanges.Length} html edits, so it can fill in the rest");
+        var remoteResult = await _remoteServiceInvoker.TryInvokeAsync<IRemoteFormattingService, ImmutableArray<TextChange>>(
+            razorDocument.Project.Solution,
+            (service, solutionInfo, cancellationToken) => service.GetOnTypeFormattingEditsAsync(solutionInfo, razorDocument.Id, htmlChanges, request.Position.ToLinePosition(), request.Character, options, cancellationToken),
+            cancellationToken).ConfigureAwait(false);
+
+        if (remoteResult is [_, ..] allChanges)
+        {
+            _logger.LogDebug($"Got a total of {allChanges.Length} ranges back from OOP");
+
+            return allChanges.Select(sourceText.GetTextEdit).ToArray();
+        }
+
+        return null;
+    }
+
+    private async Task<TextEdit[]?> GetHtmlFormattingEditsAsync(DocumentOnTypeFormattingParams request, TextDocument razorDocument, CancellationToken cancellationToken)
```

Comment:
Consider calling this `TryGetHtmlFormattingEditsAsync`

Reply:
or not :smile:

Reply:
Sorry, guess I missed this one. Added to the list to fix in the next one :)